libraries/WiFiClientSecure/src/ssl_client.cpp

```diff
@@ -391,14 +398,60 @@ int send_ssl_data(sslclient_context *ssl_client, const uint8_t *data, size_t len
     return ret;
 }
 
-int get_ssl_receive(sslclient_context *ssl_client, uint8_t *data, int length)
+// Some protocols, such as SMTP, XMPP, MySQL/Posgress and various others
+// do a 'in-line' upgrade from plaintext to SSL or TLS (usually with some
+// sort of 'STARTTLS' textual command from client to sever). For this
+// we need to have access to the 'raw' socket; i.e. without TLS/SSL state
+// handling before the handshake starts; but after setting up the TLS
+// connection.
+//
+int peek_net_receive(sslclient_context *ssl_client, int timeout) {
+#if MBEDTLS_FIXED_LINKING_NET_POLL
+    int ret = mbedtls_net_poll((mbedtls_net_context*)ssl_client, MBEDTLS_NET_POLL_READ, timeout);
+    ret == MBEDTLS_NET_POLL_READ ? 1 : ret;
+#else
+    // We should be using mbedtls_net_poll(); which is part of mbedtls and
+    // included in the EspressifSDK. Unfortunately - it did not make it into
+    // the statically linked library file. So, for now, we replace it by
+    // substancially similar code.
+    //
+    struct timeval tv = { .tv_sec = timeout / 1000, .tv_usec = (timeout % 1000) * 1000 };
+
+    fd_set fdset;
+    FD_SET(ssl_client->socket, &fdset);
+
+    int ret = select(ssl_client->socket + 1, &fdset, nullptr, nullptr, timeout<0 ? nullptr : &tv);
+    if (ret < 0) {
+        log_e("select on read fd %d, errno: %d, \"%s\"", ssl_client->socket, errno, strerror(errno));
+        lwip_close(ssl_client->socket);
+        ssl_client->socket = -1;
+        return -1;
+    };
+#endif
+    return ret;
+};
+
+int get_net_receive(sslclient_context *ssl_client, uint8_t *data, int length)
 {
-    //log_d( "Reading HTTP response...");   //for low level debug
-    int ret = -1;
+   int ret = peek_net_receive(ssl_client,ssl_client->socket_timeout);
+   if (ret > 0)
+	ret = mbedtls_net_recv(ssl_client, data, length);
 
-    ret = mbedtls_ssl_read(&ssl_client->ssl_ctx, data, length);
+    // log_v( "%d bytes NET read of %d", ret, length);   //for low level debug
+    return ret;
+}
```

Comment:
Please fix indentation

Reply:
Is there an .ident.pro file or similar that has this projects prefered formatting ?

Reply:
Not yet, but we try to keep our sources on spaces 2 or 4